modules/javafx.base/src/main/java/com/sun/javafx/binding/ConditionalBinding.java

```diff
@@ -0,0 +1,67 @@
+package com.sun.javafx.binding;
+
+import java.util.Objects;
+
+import javafx.beans.value.ObservableValue;
+
+public class ConditionalBinding<T> extends LazyObjectBinding<T> {
+
+    private final ObservableValue<T> source;
+    private final ObservableValue<Boolean> nonNullCondition;
+
+    private Subscription subscription;
+
+    public ConditionalBinding(ObservableValue<T> source, ObservableValue<Boolean> condition) {
+        this.source = Objects.requireNonNull(source, "source cannot be null");
+        this.nonNullCondition = Objects.requireNonNull(condition, "condition cannot be null").orElse(false);
+
+        // condition is always observed and never unsubscribed
+        Subscription.subscribe(nonNullCondition, current -> {
```

Comment:
do I understand this correctly - if the transient Nodes are created and hooked up with ConditionalBinding and then discarded, we'll have a memory leak?

Reply:
I'm not sure I fully understand what you mean here with the transient nodes, so if you have an example where you think this doesn't work as you'd like, I'd be happy to take a look at it.  However, I'll try explain though what's going on as best I can:

There are two important things at play here:

1) The bindings resulting from `map`, `when`, etc. are lazy bindings because they're created with the fluent system.  This means that if they're not observed themselves, they remove their listeners to their source.  For example, a chain of values (with arrows pointing from referrer to referent):

          A <-- B <-- C <-- D

Or in Java:

          b = a.map(...); c = b.map(...); d = c.map(...);

Here B observes A, C observes B and D observes C.  These will only create listeners on their source when they are observed themselves.  In their unobserved state, D can be GC'd, then C, then B, then A. When the listeners are present, the chain looks like:

         A <-> B <-> C <-> D <----> something is observing D

Now nothing can be GC'd as long as whatever is observing D is strongly referenced. However, as soon as the observer on D disappears, the chain reverts back to single directional pointers because all listeners are removed when unobserved.

2) The condition for `when` acts as an additional requirement for an observable value to create a listener on its source.  If it is `false`, no listener on its source is created, even when observed itself.  If it is `true`, then a listener is only created on its source if it is observed itself, otherwise still no listener is created.

This means when you have:

      ObservableValue<Boolean> active = new SimpleBooleanProperty(true);
 
And then use it as the condition for another property:

      ObservableValue<String> x = label.textProperty().when(active);

      x.addListener(xyz);

Then the `ObservableValue` that results from the call to `when` ("x") will always be subscribed to the `active` value. That means as long as `active` is referred, that `x` cannot be GC'd -- this is what you want, as when I toggle `active` back and forth, the thing we're toggling should not disappear.

In this case, there is a listener AND the condition is `true`, so label's text property is referencing `x` as `x` will have added a listener to label's text property.

         label.text <-> x <-----> listener
         x <-> active

If the condition is `false`, the graph instead looks like:

         label.text <-- x <-----> listener
         x <-> active

This is a very subtle difference, but it does mean that the `label` no longer has any influence on keeping these objects alive.  If only `label` is strongly referenced, then toggling `active` to `false`, and then purposely removing all references to `active` would mean the system can freely remove all these observables, including the listener as there is no way you could turn the switch back on again...

If you however kept a reference to one of these (either `listener`, `x` or `active`) then they all are referenced, as they should be, as I could toggle `active` back to `true` and expect the whole chain to work again as before.

The fact that the lifecycle of the condition and its resulting observable value are tied to each other should not matter, and in fact, has to work that way or switching on/off wouldn't do what you'd expect anymore if the properties being switched could be GC'd.

In an earlier example I showed this:

    ObservableValue<Boolean> active = new SimpleBooleanProperty(true);

    MySkin() {
         getSkinnable().textProperty().when(active).addListener(this::updateSkinStuffs);
         getSkinnable().fontProperty().when(active).addListener(this::updateSkinStuffs);
         getSkinnable().wrapTextProperty().when(active).addListener(this::updateSkinStuffs);
    }

    void dispose() {
         active.set(false);  // kills all listeners/bindings/etc in a single line
    }

Assuming the Skin itself is not strongly referenced at some point, the listeners in the above example are all still referenced by the control resulting from `getSkinnable` and can't be GC'd.  Skinnable refers to its text property, the text property refers to the `when` observable, the `when` observable refers to the `updateSkinStuffs` method, and that method pointer (but also the `when` observable which refers to `active`) keeps the entire Skin and all its listeners alive.

However, if you set `active` to false, that chain is broken.  Specifically, the text property no longer references the `when` observable, because the `when` observable will unsubscribe itself.  Since there are no other references to the skin anymore, the entire skin, including the `active` observable boolean is GC'd.  The fact that `when` and `active` reference each other makes no difference.

Reply:
Since this was quite wordy, and since pictures are worth a thousand words...

Here is the situation when you have a Label with its Skin:

![image](https://user-images.githubusercontent.com/995917/195709660-0bdad680-9240-48e9-9f8e-ae4b40ab7047.png)

When that Skin is replaced, dispose is called which toggles `active` to `false`.  Furthermore, the Label will no longer reference the Skin.  This image then shows the state of the references, and which objects can be GC'd:

![image](https://user-images.githubusercontent.com/995917/195710017-413af0ec-1c0f-4329-adb5-d5c3072c0008.png)

Reply:
Thank you for explanations, @hjohn .

This *might* work for Skins, of course, though it is not exactly what is needed for Skins (I will explain why in https://github.com/openjdk/jfx/pull/908 because of context).

In the context of this PR, I can see how a memory leak can be created when a developer has the `active` property as long lived and that would prevent the whole chain from being collected.  An example would be trying to attach to the Window.showingProperty(), adding listeners that reference Nodes that are supposed to be GC'd when removed from the said window (dynamically creating detail panes as response to selection event, for example).

I think what you propose here is good, but, as any other place where listeners are added, must be used with caution (may be reflect that fact in javadoc?)

Reply:
Also, using your example above, it might be too easy to create a memory leak by inadvertently referring to a long lived object:

```
getSkinnable().textProperty().when(active).addListener((src,prev,current) -> {
  getSkinnable().setSomething(window.getSomething() + "..."); // window reference makes the listener lambda uncollectable, resulting in a memory leak
});
```

Reply:
> Also, using your example above, it might be too easy to create a memory leak by inadvertently referring to a long lived object:
> 
> ```
> getSkinnable().textProperty().when(active).addListener((src,prev,current) -> {
>   getSkinnable().setSomething(window.getSomething() + "..."); // window reference makes the listener lambda uncollectable, resulting in a memory leak
> });
> ```

I'm not sure I follow your meaning.  Why would referring to `window` make the lambda uncollectable? The lambda is referring to `window`, not the other way around.

The current "alternative", which is weak listeners, is not what it is cracked up to be.  A weak listener that is not strongly referenced will be collected at a random time, including never, (almost) immediately or in 5 minutes.  All that time, the listener still functions, responds to potential events and property changes, even though the reason for its existence has since disappeared.

When a weak reference is collected depends on many things, including the chosen GC, its parameters, JDK versions, memory pressure, etc.  This proposal is specifically tailored to make the process of removing listeners deterministic without having to manually unregister listeners.  Weak listeners are completely unpredictable and could potentially live much longer than expected on certain GC's (or future improvements of GC's) that have different strategies for cleaning weak references.

A weak reference specifically is NOT something that gets cleaned up immediately upon the last reference disappearing. This can cause all kinds of phantom effects in JavaFX applications, if your listener does something like persisting state or printing something to the console. For example, a long lived model gets a weak listener attached to it when a dialog opens; the listener persists the last selected item.  The dialog is closed and reopened.  Now there might be two listeners, both persisting the last selected item. This may show up on your console as two database calls where one is expected.

You can argue this is bad programming and that those listeners should have been cleaned up specifically with a dispose method, and I agree, you get predictable behavior that way. And that's exactly what this proposal also brings, predictable behavior, using a syntax that doesn't require you to keep track of your listeners explicitely:

       longLivedModel.indexProperty().when(dialog::shownProperty).addListener(persistPosition);

This detaches the listener immediately after the dialog is closed.  Not in 3 seconds or 5 minutes, but as soon as it is no longer visible.  Opening a new dialog will not reinstate this listener. Reusing the same dialog however would, and it would start working again exactly as you'd expect as soon as the dialog is visible again.

Reply:
let me try to create a unit test (and I could be wrong, of course).

Reply:
> In the context of this PR, I can see how a memory leak can be created when a developer has the `active` property as long lived and that would prevent the whole chain from being collected. An example would be trying to attach to the Window.showingProperty(), adding listeners that reference Nodes that are supposed to be GC'd when removed from the said window (dynamically creating detail panes as response to selection event, for example).

What would be the purpose of having a long lived `active` property?  If it has any purpose, then the properties it is controlling must stay referenced, so yes, that would prevent the chain from being GC'd otherwise it could suddenly break your application.

Just like this application breaks currently because of weak listeners:

         label.textProperty().concat("A").addListener((src, old, current) -> System.out.println(current));

The above code will stop working at a random time due to the weak listener created by `concat`.  The same code without `concat` will keep working... this is very hard to explain to users.

Now, if you had used `map`, this wouldn't happen.  This is because `map` uses a normal listener, since it is observed itself.  If the above is what the user wants to do, then who are we to break this when a GC happens?

You also mentioned the `Window.showingProperty`. You have to use the correct path to access this property.  You don't want to make it depend on whether the window is showing.  No, you want to make it depend on whether the label has a scene which is part of a showing window:

       ObservableValue<Boolean> showing = label.flatMap(Node::sceneProperty())
             .flatMap(Scene::windowProperty)
             .orElse(false);

Or short version:

       ObservableValue<Boolean> showing = label.shownProperty();

Now, the listener on your label is added like this:

       label.textProperty().when(showing).addListener( xyz );

Or inlined:

       label.textProperty().when(label::shownProperty).addListener( xyz );

When you remove this label, it's Scene is cleared. `showing` will therefore become `false`, and in turn the listener will stop functioning.

Reply:
sorry for including multiple topics in one comment.

1. it looks like your proposal will work, as long as the `active` property is not referenced outside of its owner (in our discussion, Skin).  If that is true, as soon as you set active=false, the listener is disconnected and the skin, the active property, and the lambda can be collected.  Thus, as you correctly explained, we need to create a large aggregate with two flat maps in order to avoid the memory leak (whether you hide this complexity behind shownProperty is irrelevant).  So yes, it will work, as long as the developer makes no mistakes and does not wire the thing directly (a mistake I readily made)

2. as a side note, I would discourage a pattern where Nodes are reused and need to be reconnected.  at least in my applications, I never do that, but I can see situation when this might be useful.

3. is when() a good name?  it sort of implies a time-domain criterion instead of when a boolean becomes true (whenTrue? whenAllowed?)  i could be wrong here.

Reply:
> 1. it looks like your proposal will work, as long as the `active` property is not referenced outside of its owner (in our discussion, Skin).  If that is true, as soon as you set active=false, the listener is disconnected and the skin, the active property, and the lambda can be collected.  Thus, as you correctly explained, we need to create a large aggregate with two flat maps in order to avoid the memory leak (whether you hide this complexity behind shownProperty is irrelevant).

Yes, the `shownProperty` is there purely for convenience, and more useful for regular controls, not the skin scenario I think.  I have this helper that I use for this purpose at the moment:

    public static ObservableValue<Boolean> showing(Node node) {
      return node.sceneProperty()
        .flatMap(Scene::windowProperty)
        .flatMap(Window::showingProperty)
        .orElse(false);
    }

>  So yes, it will work, as long as the developer makes no mistakes and does not wire the thing directly (a mistake I readily made)

Yes, that's certainly true, I think the `shownProperty` will help with that though, as I think it will prevent people from storing the flatMap aggregate and reusing it, and perhaps then reusing it for the wrong controls.  Reuse however can still be fine for groups of listeners of several controls that go together.  If your control's lifecycle is tied to a group, container or dialog, then it is perfectly fine to use one of their shown properties.

> 2. as a side note, I would discourage a pattern where Nodes are reused and need to be reconnected.  at least in my applications, I never do that, but I can see situation when this might be useful.

I'm not in favor of that either, and I don't think I ever do it, but it is allowed by JavaFX, and so if you did, then the listeners would be restored as they were, and since you had to have a reference to this reused control still, nothing will have been GC'd yet.

> 3. is when() a good name?  it sort of implies a time-domain criterion instead of when a boolean becomes true (whenTrue? whenAllowed?)  i could be wrong here.

ReactFX used `conditionOn` and had a special version `conditionOnShowing` which accepts a `Node` (this would however create a circular reference between projects base and graphics).  I proposed `when` as it is nice and short, inspired by the recent developments in the area of switch expressions.  Even better would be `while` IMHO, but that is unfortunately a reserved keyword.  Still, when works reasonably well: "Listen to changes of this long lived property **when** this condition holds".  While would definitely be better or perhaps "as long as" or "whenever" :)

I don't think we should add "true" in the name, no other conditionals do this (like `Stream#filter` or `List#removeIf`).

`filter` itself is also not an option, as this has a different meaning which we may implement in the future. `filter` would allow you to remove certain values, which would set the value to empty:

       textProperty.filter(text -> !isRudeWord()).orElse("<censored>").addListener(...);

Reply:
thank you for clarifications.

If I were to choose, I'd pick `conditionOn` as it implies a boolean.

Also, since I've fallen into this trap when reviewing this PR, I think it might be a good idea to explain how to avoid memory leak by using a local property in `conditionOn/when` so as not to create a memory leak.

Reply:
@andy-goryachev-oracle 

About avoiding the memory leak.  The documentation for `when` currently says:

> The returned {@code ObservableValue} only observes this value when {@code condition} holds {@code true}. **This allows this {@code ObservableValue} and the conditional {@code ObservableValue} to be garbage collected if neither is otherwise strongly referenced when {@code condition} holds {@code false}.** This is in contrast to the general behavior of bindings, where the binding is only eligible for garbage collection when not observed itself.

The second sentence I think explains the reason that it should be a property with a similar lifecycle.

Do you think I should expand on this further?